File: VoxtralTTS/Sources/VoxtralTTS/Model/TransformerDecoder.swift
import MLX
import MLXFast
import MLXNN
import Foundation

// MARK: - RoPE Helpers

func precomputeFreqsCis(dim: Int, end: Int, theta: Float = 1_000_000.0) -> MLXArray {
    let freqs = 1.0 / MLX.pow(theta, MLXArray(stride(from: 0, to: Float(dim), by: 2)) / Float(dim))
    let t = MLXArray(Array(0 ..< end).map { Float($0) })
    let freqsGrid = MLX.outer(t, freqs)
    return MLX.stacked([MLX.cos(freqsGrid), MLX.sin(freqsGrid)], axis: -1)
}

func applyRotaryEmb(_ xq: MLXArray, _ xk: MLXArray, freqsCis: MLXArray) -> (MLXArray, MLXArray) {
    let shape = xq.shape
    let xqR = xq.reshaped(shape[0], shape[1], shape[2], -1, 2)
    let xkR = xk.reshaped(xk.shape[0], xk.shape[1], xk.shape[2], -1, 2)

    // freqs_cis shape is [T, head_dim/2, 2]
    // cos/sin: [T, head_dim/2] -> broadcast to [1, T, 1, head_dim/2]
    let cosB = freqsCis[0..., 0..., 0].reshaped(1, -1, 1, freqsCis.dim(1))
    let sinB = freqsCis[0..., 0..., 1].reshaped(1, -1, 1, freqsCis.dim(1))

    let xqOutR = xqR[0..., 0..., 0..., 0..., 0] * cosB - xqR[0..., 0..., 0..., 0..., 1] * sinB
    let xqOutI = xqR[0..., 0..., 0..., 0..., 0] * sinB + xqR[0..., 0..., 0..., 0..., 1] * cosB
    let xqOut = MLX.stacked([xqOutR, xqOutI], axis: -1).reshaped(shape)

    let xkShape = xk.shape
    let xkOutR = xkR[0..., 0..., 0..., 0..., 0] * cosB - xkR[0..., 0..., 0..., 0..., 1] * sinB
    let xkOutI = xkR[0..., 0..., 0..., 0..., 0] * sinB + xkR[0..., 0..., 0..., 0..., 1] * cosB
    let xkOut = MLX.stacked([xkOutR, xkOutI], axis: -1).reshaped(xkShape)

    return (xqOut, xkOut)
}

func repeatKv(_ x: MLXArray, nRep: Int) -> MLXArray {
    if nRep == 1 { return x }
    let (B, T, nKvHeads, headDim) = (x.dim(0), x.dim(1), x.dim(2), x.dim(3))
    let expanded = MLX.expandedDimensions(x, axis: 3)
    let broadcasted = MLX.broadcast(expanded, to: [B, T, nKvHeads, nRep, headDim])
    return broadcasted.reshaped(B, T, nKvHeads * nRep, headDim)
}

// MARK: - Attention

class Attention: Module {
    @ModuleInfo(key: "wq") var wq: Linear
    @ModuleInfo(key: "wk") var wk: Linear
    @ModuleInfo(key: "wv") var wv: Linear
    @ModuleInfo(key: "wo") var wo: Linear

    let nHeads: Int
    let nKvHeads: Int
    let headDim: Int
    let nRep: Int
    let scale: Float

    init(dim: Int, nHeads: Int, nKvHeads: Int, headDim: Int) {
        self.nHeads = nHeads
        self.nKvHeads = nKvHeads
        self.headDim = headDim
        self.nRep = nHeads / nKvHeads
        self.scale = pow(Float(headDim), -0.5)

        self._wq.wrappedValue = Linear(dim, nHeads * headDim, bias: false)
        self._wk.wrappedValue = Linear(dim, nKvHeads * headDim, bias: false)
        self._wv.wrappedValue = Linear(dim, nKvHeads * headDim, bias: false)
        self._wo.wrappedValue = Linear(nHeads * headDim, dim, bias: false)
    }

    func callAsFunction(
        _ x: MLXArray,
        freqsCis: MLXArray,
        mask: MLXArray? = nil,
        cache: (MLXArray, MLXArray)? = nil
    ) -> (MLXArray, (MLXArray, MLXArray)) {
        let B = x.dim(0)
        let T = x.dim(1)

        var q = wq(x).reshaped(B, T, nHeads, headDim)
        var k = wk(x).reshaped(B, T, nKvHeads, headDim)
        var v = wv(x).reshaped(B, T, nKvHeads, headDim)

        (q, k) = applyRotaryEmb(q, k, freqsCis: freqsCis)

        if let (kCache, vCache) = cache {
            k = MLX.concatenated([kCache, k], axis: 1)
            v = MLX.concatenated([vCache, v], axis: 1)
        }

        let newCache = (k, v)

        let kExpanded = repeatKv(k, nRep: nRep)
        let vExpanded = repeatKv(v, nRep: nRep)

        // [B, T, H, D] -> [B, H, T, D]
        let qT = q.transposed(0, 2, 1, 3)
        let kT = kExpanded.transposed(0, 2, 1, 3)
        let vT = vExpanded.transposed(0, 2, 1, 3)

        var scores = (qT.matmul(kT.transposed(0, 1, 3, 2))) * scale
        if let mask = mask {
            scores = scores + mask
        }

        let weights = softmax(scores, axis: -1)
        let output = weights.matmul(vT)
        let outputReshaped = output.transposed(0, 2, 1, 3).reshaped(B, T, -1)

        return (wo(outputReshaped), newCache)
    }
}

// MARK: - FeedForward (SwiGLU)

class FeedForward: Module {
    @ModuleInfo(key: "w1") var w1: Linear
    @ModuleInfo(key: "w2") var w2: Linear
    @ModuleInfo(key: "w3") var w3: Linear

    init(dim: Int, hiddenDim: Int) {
        self._w1.wrappedValue = Linear(dim, hiddenDim, bias: false)
        self._w2.wrappedValue = Linear(hiddenDim, dim, bias: false)
        self._w3.wrappedValue = Linear(dim, hiddenDim, bias: false)
    }

    func callAsFunction(_ x: MLXArray) -> MLXArray {
        w2(silu(w1(x)) * w3(x))
    }
}

// MARK: - Transformer Block

class TransformerBlock: Module {
    @ModuleInfo(key: "attention") var attention: Attention
    @ModuleInfo(key: "feed_forward") var feedForward: FeedForward
    @ModuleInfo(key: "attention_norm") var attentionNorm: RMSNorm
    @ModuleInfo(key: "ffn_norm") var ffnNorm: RMSNorm

    init(dim: Int, nHeads: Int, nKvHeads: Int, headDim: Int, hiddenDim: Int, normEps: Float) {
        self._attention.wrappedValue = Attention(dim: dim, nHeads: nHeads, nKvHeads: nKvHeads, headDim: headDim)
        self._feedForward.wrappedValue = FeedForward(dim: dim, hiddenDim: hiddenDim)
        self._attentionNorm.wrappedValue = RMSNorm(dimensions: dim, eps: normEps)
        self._ffnNorm.wrappedValue = RMSNorm(dimensions: dim, eps: normEps)
    }

    func callAsFunction(
        _ x: MLXArray,
        freqsCis: MLXArray,
        mask: MLXArray? = nil,
        cache: (MLXArray, MLXArray)? = nil
    ) -> (MLXArray, (MLXArray, MLXArray)) {
        let (h, newCache) = attention(attentionNorm(x), freqsCis: freqsCis, mask: mask, cache: cache)
        let x2 = x + h
        let x3 = x2 + feedForward(ffnNorm(x2))
        return (x3, newCache)
    }
}

// MARK: - Mistral Transformer Decoder

class MistralTransformerDecoder: Module {
    @ModuleInfo(key: "tok_embeddings") var tokEmbeddings: Embedding
    @ModuleInfo(key: "layers") var layers: [TransformerBlock]
    @ModuleInfo(key: "norm") var norm: RMSNorm

    // Optional output projection (only when not tying embeddings)
    @ModuleInfo(key: "output") var output: Linear?

    let dim: Int
    let vocabSize: Int
    let ropeTheta: Float
    let tieWordEmbeddings: Bool

    init(config: ModelConfig) {
        self.dim = config.dim
        self.vocabSize = config.vocabSize
        self.ropeTheta = config.ropeTheta
        self.tieWordEmbeddings = config.tieWordEmbeddings

        self._tokEmbeddings.wrappedValue = Embedding(embeddingCount: config.vocabSize, dimensions: config.dim)
        self._layers.wrappedValue = (0 ..< config.nLayers).map { _ in
            TransformerBlock(
                dim: config.dim,
                nHeads: config.nHeads,
                nKvHeads: config.nKvHeads,
                headDim: config.headDim,
                hiddenDim: config.hiddenDim,
                normEps: config.normEps
            )
        }
        self._norm.wrappedValue = RMSNorm(dimensions: config.dim, eps: config.normEps)

        if !config.tieWordEmbeddings {
            self._output.wrappedValue = Linear(config.dim, config.vocabSize, bias: false)
        }
    }

    func callAsFunction(
        tokens: MLXArray? = nil,
        cache: [(MLXArray, MLXArray)?]? = nil,
        inputEmbeds: MLXArray? = nil
    ) -> (logits: MLXArray, hiddenStates: MLXArray, newCache: [(MLXArray, MLXArray)]) {
        var h: MLXArray
        let T: Int

        if let embeds = inputEmbeds {
            h = embeds
            T = embeds.dim(1)
        } else {
            guard let tokens = tokens else {
                fatalError("Either tokens or inputEmbeds must be provided")
            }
            T = tokens.dim(1)
            h = tokEmbeddings(tokens)
        }

        // Position offset from cache
        var offset = 0
        if let c = cache, let first = c[0] {
            offset = first.0.dim(1)
        }

        // Precompute RoPE
        let headDim = layers[0].attention.headDim
        let allFreqs = precomputeFreqsCis(dim: headDim, end: offset + T, theta: ropeTheta)
        let freqsCis = allFreqs[offset ..< (offset + T)]

        // Causal mask
        var mask: MLXArray? = nil
        if T > 1 {
            mask = MultiHeadAttention.createAdditiveCausalMask(T).asType(h.dtype)
        }

        var newCache: [(MLXArray, MLXArray)] = []
        for (i, layer) in layers.enumerated() {
            let layerCache = cache?[i]
            let (hNew, c) = layer(h, freqsCis: freqsCis, mask: mask, cache: layerCache)
            h = hNew
            newCache.append(c)
        }

        h = norm(h)
        let hiddenStates = h  // Post-norm hidden states for downstream use

        let logits: MLXArray
        if tieWordEmbeddings {
            logits = h.matmul(tokEmbeddings.weight.T)
        } else {
            logits = output!(h)
        }

        return (logits, hiddenStates, newCache)
    }
}

File: voxtral_tts/transformer_decoder.py
"""Mistral-based Transformer Decoder for Voxtral TTS.

This is the text LLM component (3.4B params) that generates hidden states
from text tokens. Based on Ministral-3B architecture.
"""

import math
from typing import Optional, Tuple

import mlx.core as mx
import mlx.nn as nn


class RMSNorm(nn.Module):
    def __init__(self, dim: int, eps: float = 1e-5):
        super().__init__()
        self.weight = mx.ones((dim,))
        self.eps = eps

    def __call__(self, x: mx.array) -> mx.array:
        return mx.fast.rms_norm(x, self.weight, self.eps)


def precompute_freqs_cis(dim: int, end: int, theta: float = 1000000.0) -> mx.array:
    freqs = 1.0 / (theta ** (mx.arange(0, dim, 2, dtype=mx.float32) / dim))
    t = mx.arange(end, dtype=mx.float32)
    freqs = mx.outer(t, freqs)
    return mx.stack([mx.cos(freqs), mx.sin(freqs)], axis=-1)


def apply_rotary_emb(xq: mx.array, xk: mx.array, freqs_cis: mx.array) -> Tuple[mx.array, mx.array]:
    """Apply rotary position embeddings to query and key tensors."""
    # xq, xk: [B, T, n_heads, head_dim]
    # freqs_cis: [T, head_dim/2, 2]
    xq_r = xq.reshape(*xq.shape[:-1], -1, 2)  # [B, T, H, head_dim/2, 2]
    xk_r = xk.reshape(*xk.shape[:-1], -1, 2)

    # Extract cos, sin from freqs_cis
    cos = freqs_cis[:, :, 0]  # [T, head_dim/2]
    sin = freqs_cis[:, :, 1]

    # Broadcast: [1, T, 1, head_dim/2]
    cos = cos[None, :, None, :]
    sin = sin[None, :, None, :]

    # Apply rotation
    xq_out_r = xq_r[..., 0] * cos - xq_r[..., 1] * sin
    xq_out_i = xq_r[..., 0] * sin + xq_r[..., 1] * cos
    xq_out = mx.stack([xq_out_r, xq_out_i], axis=-1).reshape(xq.shape)

    xk_out_r = xk_r[..., 0] * cos - xk_r[..., 1] * sin
    xk_out_i = xk_r[..., 0] * sin + xk_r[..., 1] * cos
    xk_out = mx.stack([xk_out_r, xk_out_i], axis=-1).reshape(xk.shape)

    return xq_out, xk_out


def repeat_kv(x: mx.array, n_rep: int) -> mx.array:
    """Repeat KV heads to match number of query heads."""
    if n_rep == 1:
        return x
    B, T, n_kv_heads, head_dim = x.shape
    x = mx.expand_dims(x, axis=3)  # [B, T, n_kv_heads, 1, head_dim]
    x = mx.broadcast_to(x, (B, T, n_kv_heads, n_rep, head_dim))
    return x.reshape(B, T, n_kv_heads * n_rep, head_dim)


class Attention(nn.Module):
    def __init__(self, dim: int, n_heads: int, n_kv_heads: int, head_dim: int):
        super().__init__()
        self.n_heads = n_heads
        self.n_kv_heads = n_kv_heads
        self.head_dim = head_dim
        self.n_rep = n_heads // n_kv_heads
        self.scale = head_dim ** -0.5

        self.wq = nn.Linear(dim, n_heads * head_dim, bias=False)
        self.wk = nn.Linear(dim, n_kv_heads * head_dim, bias=False)
        self.wv = nn.Linear(dim, n_kv_heads * head_dim, bias=False)
        self.wo = nn.Linear(n_heads * head_dim, dim, bias=False)

    def __call__(
        self,
        x: mx.array,
        freqs_cis: mx.array,
        mask: Optional[mx.array] = None,
        cache: Optional[Tuple[mx.array, mx.array]] = None,
    ) -> Tuple[mx.array, Optional[Tuple[mx.array, mx.array]]]:
        B, T, _ = x.shape

        q = self.wq(x).reshape(B, T, self.n_heads, self.head_dim)
        k = self.wk(x).reshape(B, T, self.n_kv_heads, self.head_dim)
        v = self.wv(x).reshape(B, T, self.n_kv_heads, self.head_dim)

        q, k = apply_rotary_emb(q, k, freqs_cis)

        if cache is not None:
            k_cache, v_cache = cache
            k = mx.concatenate([k_cache, k], axis=1)
            v = mx.concatenate([v_cache, v], axis=1)

        new_cache = (k, v)

        # GQA: repeat KV heads
        k = repeat_kv(k, self.n_rep)
        v = repeat_kv(v, self.n_rep)

        # [B, T, H, D] -> [B, H, T, D]
        q = q.transpose(0, 2, 1, 3)
        k = k.transpose(0, 2, 1, 3)
        v = v.transpose(0, 2, 1, 3)

        scores = (q @ k.transpose(0, 1, 3, 2)) * self.scale

        if mask is not None:
            scores = scores + mask

        weights = mx.softmax(scores, axis=-1)
        output = weights @ v  # [B, H, T, D]
        output = output.transpose(0, 2, 1, 3).reshape(B, T, -1)

        return self.wo(output), new_cache


class FeedForward(nn.Module):
    def __init__(self, dim: int, hidden_dim: int):
        super().__init__()
        self.w1 = nn.Linear(dim, hidden_dim, bias=False)
        self.w2 = nn.Linear(hidden_dim, dim, bias=False)
        self.w3 = nn.Linear(dim, hidden_dim, bias=False)

    def __call__(self, x: mx.array) -> mx.array:
        return self.w2(nn.silu(self.w1(x)) * self.w3(x))


class TransformerBlock(nn.Module):
    def __init__(self, dim: int, n_heads: int, n_kv_heads: int,
                 head_dim: int, hidden_dim: int, norm_eps: float = 1e-5):
        super().__init__()
        self.attention = Attention(dim, n_heads, n_kv_heads, head_dim)
        self.feed_forward = FeedForward(dim, hidden_dim)
        self.attention_norm = RMSNorm(dim, eps=norm_eps)
        self.ffn_norm = RMSNorm(dim, eps=norm_eps)

    def __call__(
        self,
        x: mx.array,
        freqs_cis: mx.array,
        mask: Optional[mx.array] = None,
        cache: Optional[Tuple[mx.array, mx.array]] = None,
    ) -> Tuple[mx.array, Optional[Tuple[mx.array, mx.array]]]:
        h, new_cache = self.attention(self.attention_norm(x), freqs_cis, mask, cache)
        x = x + h
        x = x + self.feed_forward(self.ffn_norm(x))
        return x, new_cache


class MistralTransformerDecoder(nn.Module):
    """Mistral-based transformer decoder for text-to-hidden-state generation.

    Architecture: Ministral-3B variant with 26 layers, dim=3072.
    """

    def __init__(
        self,
        dim: int = 3072,
        n_layers: int = 26,
        n_heads: int = 32,
        n_kv_heads: int = 8,
        head_dim: int = 128,
        hidden_dim: int = 9216,
        vocab_size: int = 131072,
        norm_eps: float = 1e-5,
        rope_theta: float = 1000000.0,
        max_position_embeddings: int = 128000,
        tie_word_embeddings: bool = True,
    ):
        super().__init__()
        self.dim = dim
        self.n_layers = n_layers
        self.vocab_size = vocab_size
        self.rope_theta = rope_theta
        self.max_position_embeddings = max_position_embeddings
        self.tie_word_embeddings = tie_word_embeddings

        self.tok_embeddings = nn.Embedding(vocab_size, dim)
        self.layers = [
            TransformerBlock(dim, n_heads, n_kv_heads, head_dim, hidden_dim, norm_eps)
            for _ in range(n_layers)
        ]
        self.norm = RMSNorm(dim, eps=norm_eps)

        if not tie_word_embeddings:
            self.output = nn.Linear(dim, vocab_size, bias=False)

    def __call__(
        self,
        tokens: Optional[mx.array] = None,
        cache: Optional[list] = None,
        input_embeds: Optional[mx.array] = None,
    ) -> Tuple[mx.array, mx.array, Optional[list]]:
        """Forward pass through the decoder.

        Args:
            tokens: Input token IDs [B, T] (mutually exclusive with input_embeds)
            cache: Optional KV cache list
            input_embeds: Pre-computed embeddings [B, T, dim] to use instead of
                token lookup. Used for feeding audio embeddings back into the LLM.

        Returns:
            logits: [B, T, vocab_size]
            hidden_states: [B, T, dim] (last layer hidden states)
            new_cache: Updated KV cache
        """
        if input_embeds is not None:
            h = input_embeds
            B, T = h.shape[:2]
        else:
            B, T = tokens.shape
            h = self.tok_embeddings(tokens)

        # Compute position offset from cache
        offset = 0
        if cache is not None and cache[0] is not None:
            offset = cache[0][0].shape[1]

        # Precompute RoPE frequencies
        freqs_cis = precompute_freqs_cis(
            self.layers[0].attention.head_dim,
            offset + T,
            self.rope_theta,
        )[offset:offset + T]

        # Causal mask
        mask = None
        if T > 1:
            mask = nn.MultiHeadAttention.create_additive_causal_mask(T)
            mask = mask.astype(h.dtype)

        new_cache = []
        for i, layer in enumerate(self.layers):
            layer_cache = cache[i] if cache is not None else None
            h, c = layer(h, freqs_cis, mask, layer_cache)
            new_cache.append(c)

        h = self.norm(h)
        hidden_states = h  # Post-norm hidden states for downstream use

        if self.tie_word_embeddings:
            logits = self.tok_embeddings.as_linear(h)
        else:
            logits = self.output(h)

        return logits, hidden_states, new_cache
